botocore/compress.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import gzip
+import io
+import logging
+
+from botocore.utils import determine_content_length
+
+logger = logging.getLogger(__name__)
+
+
+class RequestCompressor:
+    """A class that can compress a request body."""
+
+    @classmethod
+    def compress(cls, config, request_dict, operation_model):
+        """Compresses the request body using the specified encodings.
+
+        Check if the request should be compressed based on the contents of its
+        body and config settings. Set or append the `Content-Encoding` header
+        with the matched encoding if not present.
+        """
+        body = request_dict['body']
+        if cls._should_compress_request(config, body, operation_model):
+            encodings = operation_model.request_compression['encodings']
+            headers = request_dict['headers']
+            for encoding in encodings:
+                encoder = getattr(cls, f'_{encoding}_compress_body', None)
+                if encoder is not None:
+                    ce_header = headers.get('Content-Encoding')
+                    if ce_header is None:
+                        headers['Content-Encoding'] = encoding
+                    elif encoding not in ce_header.split(','):
+                        headers['Content-Encoding'] = f'{ce_header},{encoding}'
+                    logger.debug(
+                        'Compressing request with %s encoding', encoding
+                    )
+                    request_dict['body'] = encoder(body)
+                else:
+                    logger.debug(
+                        'Unsupported compression encoding: %s' % encoding
+                    )
```

Comment:
I think we discussed this earlier, but doesn't this implementation violate the design? You're encoding the body with _all_ encodings that match an implemented function. Was there a reason we didn't fix this? We should be selecting the first supported and then moving on.

```suggestion
        if not cls._should_compress_request(config, body, operation_model):
            return
        encodings = operation_model.request_compression['encodings']
        encoding = _choose_encoding(encodings)
        if encoding is not None:
            encoder = encoding_map.get(encoding)
            logger.debug('Compressing request with %s encoding', encoding)
            request_dict['body'] = encoder(body)
            _set_compression_header(request_dict['headers'], encoding)
        
    def _set_compression_header(headers, encoding):
        content_encoding = headers.get('Content-Encoding')
        if content_encoding is None:
            headers['Content-Encoding'] = encoding
        elif encoding not in ce_header.split(','):
            headers['Content-Encoding'] = f'{ce_header},{encoding}'
                
    def _choose_encoding(encodings):
        for encoding in encodings:
            if encoding in encoding_map:
                return encoding
            else:
                logger.debug('Unsupported compression encoding: %s', encoding)
```

Reply:
Ah so sorry yes this was a miss on my part. There is a missing `return` statement after `request_dict['body'] = encoder(body)`. I'll add a test to ensure that compression isn't attempted more than once when there is more than one compression encoding available in the operation model.

Reply:
If you think its a bit easier to read/digest by yanking the selection of an encoding into a helper function, I can certainly do that.

Reply:
I elected to not use a `_choose_encoding` helper method and instead include a `return` statement in the for loop if the encoder isn't None.  This makes it easier to log without passing the encoding name around.